dff/stats/savers/clickhouse.py

```diff
@@ -0,0 +1,112 @@
+"""
+Clickhouse Saver
+----------------
+Provides the Clickhouse version of the :py:class:`.dff.stats.savers.saver.Saver`.
+The class should be constructed by calling the :py:func:`~dff.stats.savers.saver_factory`
+factory with specific parameters.
+
+"""
+import json
+from typing import List
+from urllib import parse
+import asyncio
+
+from pydantic import validator
+
+try:
+    from httpx import AsyncClient
+    from aiochclient import ChClient
+
+    IMPORT_ERROR_MESSAGE = None
+except ImportError as e:
+    IMPORT_ERROR_MESSAGE = e.msg
+
+from .saver import Saver
+from ..record import StatsRecord
+
+
+class CHItem(StatsRecord):
+    """
+    This class preemptively converts record data to JSON
+    for compliance with Clickhouse Driver restrictions.
+    """
+
+    timestamp: str
+    data: str
+
+    @validator("data", pre=True)
+    def val_data(cls, data):
+        if not isinstance(data, str):
+            return json.dumps(data)
+        return data
+
+    @validator("timestamp", pre=True)
+    def val_timestamp(cls, data):
+        if not isinstance(data, str):
+            return data.isoformat()
+        return data
+
+
+class ClickHouseSaver(Saver):
+    """
+    Saves and reads the stats dataframe from a Clickhouse database.
+    The class should be constructed by calling the :py:func:`~dff.stats.savers.saver_factory`
+    factory with specific parameters.
+
+    :param path: The construction path.
+        It should match the sqlalchemy :py:class:`~sqlalchemy.engine.Engine` initialization string.
+
+        .. code-block::
+
+            ClickHouseSaver("clickhouse://user:password@localhost:8000/default")
+
+    :param table: Sets the name of the db table to use. Defaults to "dff_stats".
+    """
+
+    def __init__(self, path: str, table: str = "dff_stats") -> None:
+        if IMPORT_ERROR_MESSAGE is not None:
+            raise ImportError(IMPORT_ERROR_MESSAGE)
+        self.table = table
+        parsed_path = parse.urlparse(path)
+        auth, _, address = parsed_path.netloc.partition("@")
+        self.db = parsed_path.path.strip("/")
+        self.url = parse.urlunparse(("http", address, "/", "", "", ""))
+        user, _, password = auth.partition(":")
+        http_client = AsyncClient()
+        if not all([self.db, self.url, user, password]):
+            raise ValueError("Invalid database URI or credentials")
+        self.ch_client = ChClient(http_client, url=self.url, user=user, password=password, database=self.db)
+        self._table_exists = False
```

Comment:
`_table_exists` is no longer necessary if the line right after this one sets it to `True`.